design/cstor/v1/pool_expansion_by_expanding_disk.md

```diff
@@ -0,0 +1,145 @@
+- This proposal includes design details of the pool expansion when underlying disks were expanded.
+
+## Motivation
+
+There were cases where OpenEBS users expand the pool by expanding underlying disks(which were used by pool) when space on the pool got exhausted.
+
+### Goals
+
+- Expand cStor pool when the underlying disk got expanded.
+
+### Non-Goals
+
+NA
+
+## Proposal
+
+### User Stories
+
+As a OpenEBS user I should be able to expand the pool when underlying disk got expanded.
+
+### Current Implementation
+
+When user creates the CSPC, CSPC controller which is watching for CSPC will get create request and creats cstorpoolinstances and cstorpooldeployments. CstorPoolInstance is responsible for holding the blockdevice information and other pool configurations while cStor pool deployment contains following containers cstor-pool-mgmt, cstor-pool container and m-exporter as name conveyes cstor-pool-mgmt is responsible for managing the pool. cstor-pool container creates pool(ZFS file system) on top of the blockdevices. If any changes need to make on pool configuration like adding blockdevices, adding raid groups and replacing blockdevice cstor-pool-mgmt container will look into CstorPoolInstance for pool information and verifies are there any pending actions to be performed then cstor-pool-mgmt will prforms corresponding operation on the pool. Below info is hight level info to know how the operation is detected
+
+- Pool Expansion When blockdevice/raid group added:	
+	cstor-pool-mgmt will iterate over all the blockdevices present in CstorPoolInstance and then verifies whether corresponding blockdevice is in use in the pool. If that blockdevice/raid group doesn’t exist then trigger is for pool expansion by adding blockdevice/raid group.
+
+- BlockDevice replacement:
+	cstor-pool-mgmt will iterate over all the blockdevices present in CstorPoolInstance and the verifies is that blockdevice is for replacement(We are maintaining annotation called openebs.io/predecessor: <replaced_blockdevice_name> on the claim of that blockdevice) if that block device is for replacement then the it will perform replacement operation on the pool.
+
+### Proposed Implementations
+
+CStor-pool-mgmt has CSPI controller for every sync operation cstor-pool-mgmt will iterate over all the blockdevices exist on CSPI by iterating over CstorPoolInstanceBlockDevices of all raid groups. Now if the capacity exist on blockdevice CR is greater than the  capacity exists on CstorPoolInstanceBlockDevice then we can follow the steps to perform the pool expansion. But anyways it is not good practice to perform expansion of pool with out getting the permission from user. So to get the input from the user we will be introducing a new field under CstorPoolInstanceBlockDevice called IsBlockDeviceExpanded. If  IsBlockDeviceExpanded sets true then it is something like user/admin is requesting to perform pool expansion steps. If IsBlockDeviceExpanded sets false then NoOps will be performed.
+
```

Comment:
CStor-pool Manger inbuilt CSPI controller watches for CSPI resources and triee to get the desired state via every sync operation. It will iterate over all the `BlockDevices` exist in different `CstorPoolInstanceBlockDevices` of all raid groups on a CSPI resource. If BlockDevice resource updated capacity is greater than the existing capacity of BlockDevice on `CstorPoolInstanceBlockDevice` then we can follow the below mentioned steps to perform the pool expansion

Reply:
Break the above para after above info and then mention about the `IsBlockDeviceExpanded` control flag.

Reply:
Done